src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateResolver.cs

```diff
@@ -485,5 +443,197 @@ private static void AddParameterMatchingToTemplates(IReadOnlyCollection<ITemplat
+            Dictionary<string, HashSet<string>> shortNamesByGroup = new Dictionary<string, HashSet<string>>();
+
+            // get the short names lists for the groups
+            foreach (ITemplateInfo template in templateList)
+            {
+                string effectiveGroupIdentity = !string.IsNullOrEmpty(template.GroupIdentity)
+                    ? template.GroupIdentity
+                    : template.Identity;
+
+                if (!shortNamesByGroup.TryGetValue(effectiveGroupIdentity, out HashSet<string> shortNames))
+                {
+                    shortNames = new HashSet<string>();
+                    shortNamesByGroup[effectiveGroupIdentity] = shortNames;
+                }
+
+                if (template is IShortNameList templateWithShortNameList)
+                {
+                    shortNames.UnionWith(templateWithShortNameList.ShortNameList);
+                }
+                else
+                {
+                    shortNames.Add(template.ShortName);
+                }
+            }
+
+            // create the TemplateInfoWithGroupShortNames with the group short names
+            List<TemplateInfoWithGroupShortNames> filterableTemplateList = new List<TemplateInfoWithGroupShortNames>();
+
+            foreach (ITemplateInfo template in templateList)
+            {
+                string effectiveGroupIdentity = !string.IsNullOrEmpty(template.GroupIdentity)
+                    ? template.GroupIdentity
+                    : template.Identity;
+
+                TemplateInfoWithGroupShortNames filterableTemplate = TemplateInfoWithGroupShortNames.FromITemplateInfo(template);
+                filterableTemplate.GroupShortNameList = shortNamesByGroup[effectiveGroupIdentity].ToList();
+                filterableTemplateList.Add(filterableTemplate);
+            }
+
+            return filterableTemplateList;
+        }
+
+        /// <summary>
+        /// Unlike <see cref="WellKnownSearchFilters.NameFilter(string)"/> the filter also matches other template group short names the template belongs to.
+        /// </summary>
+        /// <param name="name">the name to match with template name or short name</param>
+        /// <returns></returns>
+        private static Func<ITemplateInfo, MatchInfo?> CliNameFilter(string name)
+        {
+            return (template) =>
+            {
+                var groupAwareTemplate = template as TemplateInfoWithGroupShortNames ?? throw new ArgumentException("CliNameFilter filter supports only TemplateInfoWithGroupShortNames templates");
+
+                if (string.IsNullOrEmpty(name))
+                {
+                    return new MatchInfo { Location = MatchLocation.Name, Kind = MatchKind.Partial };
+                }
+
+                int nameIndex = groupAwareTemplate.Name.IndexOf(name, StringComparison.OrdinalIgnoreCase);
+
+                if (nameIndex == 0 && string.Equals(groupAwareTemplate.Name, name, StringComparison.OrdinalIgnoreCase))
```

Comment:
`string.Equals` can be replaced with string length comparison. Also applies to the other case below.